File: c#/money-problem.Domain/Portfolio.cs
using LanguageExt;

namespace money_problem.Domain;

public class Portfolio
{
    private readonly Seq<Money> moneys;

    public Portfolio() => this.moneys = Seq<Money>.Empty;

    private Portfolio(Seq<Money> moneys) => this.moneys = moneys;

    private List<Either<Error, Money>> GetConvertedMoneys(Bank bank, Currency currency) =>
        this.moneys
            .Select(money => bank.Convert(money, currency))
            .ToList();

    private static bool ContainsFailure(IEnumerable<Either<Error, Money>> results) =>
        results.Any(result => result.IsLeft);

    private string ToFailure(IEnumerable<Either<Error, Money>> results) =>
        $"Missing exchange rate(s): {GetMissingRates(results)}";

    private static string GetMissingRates(IEnumerable<Either<Error, Money>> missingRates) => missingRates
        .Match(_ => string.Empty, failure => $"[{failure.Message}]")
        .Where(message => !string.IsNullOrEmpty(message))
        .Aggregate((r1, r2) => $"{r1},{r2}");

    private Money ToSuccess(IEnumerable<Either<Error, Money>> results, Currency currency) =>
        new(results
                .Where(result => result.IsRight)
                .Sum(result => result.IfLeft(Money.Empty(currency)).Amount),
            currency);

    public Portfolio Add(Money money) => new(this.moneys.Add(money));

    public Either<string, Money> Evaluate(Bank bank, Currency currency)
    {
        var results = this.GetConvertedMoneys(bank, currency);
        return ContainsFailure(results)
            ? Either<string, Money>.Left(this.ToFailure(results))
            : Either<string, Money>.Right(this.ToSuccess(results, currency));
    }
}

File: scala/src/main/scala/domain/Portfolio.scala
package domain

import domain.Currency.Currency

sealed class Portfolio(private val moneys: Money*) {
  def add(money: Money): Portfolio =
    new Portfolio(moneys :+ money: _*)

  def evaluate(bank: Bank, toCurrency: Currency): Either[String, Money] = {
    val convertedMoneys = convertMoneys(bank, toCurrency)

    if (containsFailure(convertedMoneys))
      Left(toFailure(convertedMoneys))
    else Right(sumConvertedMoney(toCurrency, convertedMoneys))
  }

  private def convertMoneys(
      bank: Bank,
      toCurrency: Currency
  ): Seq[Either[String, Money]] =
    moneys.map(money => bank.convert(money, toCurrency))

  private def sumConvertedMoney(
      toCurrency: Currency,
      convertedMoneys: Seq[Either[String, Money]]
  ): Money = {
    Money(
      convertedMoneys
        .collect { case Right(x) => x }
        .foldLeft(0d)((acc, money) => acc + money.amount),
      toCurrency
    )
  }

  private def containsFailure(
      convertedMoneys: Seq[Either[String, Money]]
  ): Boolean = convertedMoneys.exists(_.isLeft)

  private def toFailure(
      convertedMoneys: Seq[Either[String, Money]]
  ): String = {
    convertedMoneys
      .collect { case Left(x) => x }
      .map(failure => s"[$failure]")
      .mkString("Missing exchange rate(s): ", ",", "")
  }
}
